### PR TITLE
ToC refactoring

### PR DESCRIPTION
**Pull request by robrix**
This PR refactors ToC summarization, setting up for a new approach to computing them accommodating both à la carte & precise terms.

### CODE REVIEW

File: src/Semantic/Api/TOCSummaries.hs [hunk 1]
```diff
@@ -23,26 +26,45 @@ diffSummaryBuilder format blobs = diffSummary blobs >>= serialize format
...
+  Changed  -> MODIFIED
+  Deleted  -> REMOVED
+  Inserted -> ADDED
+  Replaced -> MODIFIED
+
+toChange :: TOCSummary -> TOCSummaryChange
+toChange TOCSummary{..} = defMessage
+  & P.category   .~ formatKind kind
+  & P.term       .~ ident
+  & P.maybe'span .~ (converting #? span)
```
Comment (patrickt) [inline src/Semantic/Api/TOCSummaries.hs]:
```suggestion
  & P.maybe'span ?~ converting # span
```

No parens, and you save a whole character (🌌🧠, I know), by using the `?~` operator, which sets a `Maybe`-valued lens to `Just` a value. (I find this a bit cleaner than `#?`, which is a kludge of my own invention.)

File: src/Semantic/Api/TOCSummaries.hs [hunk 2]
```diff
@@ -23,26 +26,45 @@ diffSummaryBuilder format blobs = diffSummary blobs >>= serialize format
...
+  & P.changeType .~ toChangeType change
+
+toError :: ErrorSummary -> TOCSummaryError
+toError ErrorSummary{..} = defMessage
+  & P.error      .~ message
+  & P.maybe'span .~ converting #? span
```
Comment (patrickt) [inline src/Semantic/Api/TOCSummaries.hs]:
```suggestion
  & P.maybe'span ?~ converting # span
```

File: test/Rendering/TOC/Spec.hs [hunk 3]
```diff
@@ -228,16 +213,8 @@ blankDiff :: Diff'
...
-                  )
-               => Parser (Term syntax Loc)
-               -> BlobPair
-               -> m (Diff syntax (Maybe Declaration) (Maybe Declaration))
-diffWithParser parser blobs = diffTermPair . runJoin <$> distributeFor blobs (\ blob -> decoratorWithAlgebra (declarationAlgebra blob) <$> parse parser blob)
+summarize
+  :: DiffEffects sig m
+  => BlobPair
+  -> m [Either ErrorSummary TOCSummary]
+summarize = decoratingDiffWith summarizeDiffParsers decorateTerm (pure . summarizeDiff)
```
[the remark this answers is not in the document]
Comment (patrickt) [reply test/Rendering/TOC/Spec.hs]: Wow!

File: src/Rendering/TOC.hs [hunk 4]
```diff
@@ -98,54 +100,21 @@ newtype DedupeKey = DedupeKey (T.Text, T.Text) deriving (Eq, Ord)
...
-
--- | Construct a 'TOCSummary' from an 'Entry'.
-entrySummary :: Entry Declaration -> TOCSummary
-entrySummary entry = case entry of
-  Changed  a -> recordSummary "modified" a
-  Deleted  a -> recordSummary "removed" a
-  Inserted a -> recordSummary "added" a
-  Replaced a -> recordSummary "modified" a
+dedupe :: [(Change, Declaration)] -> [(Change, Declaration)]
+dedupe = map ((change :: Dedupe -> Change) &&& decl) . sortOn index . Map.elems . foldl' go Map.empty . zipWith (uncurry . Dedupe) [0..] where
```
Comment (patrickt) [inline src/Rendering/TOC.hs]: This line is a little long; I would write it with newlines between invocations of `.`, which gives you some nice space on the page to do FORTH-style comments about what happens at each step of the pipeline.

File: src/Analysis/TOCSummary.hs [hunk 5]
```diff
@@ -10,111 +13,129 @@ import Prologue hiding (project)
...
+--   This instance determines whether or not there is an instance for @syntax@ by looking it up in the 'DeclarationStrategy' type family. Thus producing a 'Declaration' for a node requires both defining a @'HasDeclarationBy' ''Custom'@ instance _and_ adding a definition for the type to the 'DeclarationStrategy' type family to return 'Custom'.
 --
 --   Note that since 'DeclarationStrategy' has a fallback case for its final entry, this instance will hold for all types of kind @* -> *@. Thus, this must be the only instance of 'HasDeclaration', as any other instance would be indistinguishable.
-instance (DeclarationStrategy syntax ~ strategy, HasDeclarationWithStrategy strategy whole syntax) => HasDeclaration' whole syntax where
-  toDeclaration' = toDeclarationWithStrategy (Proxy :: Proxy strategy)
+instance (DeclarationStrategy syntax ~ strategy, HasDeclarationBy strategy syntax) => HasDeclaration syntax where
+  toDeclaration = toDeclarationBy @strategy
```
[the remark this answers is not in the document]
Comment (robrix) [reply src/Analysis/TOCSummary.hs]: Yep. Them parameters ain’t free. I try not to let `AllowAmbiguousTypes` leak across modules, tho.

File: src/Rendering/TOC.hs [hunk 6]
```diff
@@ -1,95 +1,97 @@
...
-                  -> [Entry a]                            -- ^ A list of entries for relevant changed nodes in the diff.
+                  -> [(Change, a)]                        -- ^ A list of entries for relevant changed nodes in the diff.
 tableOfContentsBy selector = fromMaybe [] . cata (\ r -> case r of
   Patch patch -> (pure . patchEntry <$> bicrosswalk selector selector patch) <> bifoldMap fold fold patch <> Just []
   Merge (In (_, ann2) r) -> case (selector (In ann2 r), fold r) of
-    (Just a, Just entries) -> Just (Changed a : entries)
+    (Just a, Just entries) -> Just ((Changed, a) : entries)
     (_     , entries)      -> entries)
-   where patchEntry = patch Deleted Inserted (const Replaced)
+  where patchEntry = patch (Deleted,) (Inserted,) (const (Replaced,))
```
[the remark this answers is not in the document]
Comment (robrix) [reply src/Rendering/TOC.hs]: smh

File: src/Semantic/Api/TOCSummaries.hs (hunk 1 above)
Comment (robrix) [reply src/Semantic/Api/TOCSummaries.hs]: Nice 👍 Thanks for the suggestion & explanation!

**Pull request merged**